In [1]:
import os
project_name = "reco-tut-ysr"; branch = "main"; account = "sparsh-ai"
project_path = os.path.join('/content', project_name)

if not os.path.exists(project_path):
    !cp /content/drive/MyDrive/mykeys.py /content
    import mykeys
    !rm /content/mykeys.py
    path = "/content/" + project_name; 
    !mkdir "{path}"
    %cd "{path}"
    import sys; sys.path.append(path)
    !git config --global user.email "recotut@recohut.com"
    !git config --global user.name  "reco-tut"
    !git init
    !git remote add origin https://"{mykeys.git_token}":x-oauth-basic@github.com/"{account}"/"{project_name}".git
    !git pull origin "{branch}"
    !git checkout main
else:
    %cd "{project_path}"

/content/reco-tut-ysr


In [24]:
!git status

On branch main
Your branch is up to date with 'origin/main'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	data/silver/

nothing added to commit but untracked files present (use "git add" to track)


In [25]:
!git add . && git commit -m 'commit' && git push origin main

[main 6145636] commit
 2 files changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 data/silver/clean_playlist.pickle
 create mode 100644 data/silver/songs.pickle
Counting objects: 6, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (6/6), done.
Writing objects: 100% (6/6), 8.65 MiB | 2.72 MiB/s, done.
Total 6 (delta 1), reused 0 (delta 0)
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To https://github.com/sparsh-ai/reco-tut-ysr.git
   ca5e3d1..6145636  main -> main


---

In [10]:
import pandas as pd
import numpy as np
import itertools
import pickle

import warnings
warnings.filterwarnings("ignore")

In [13]:
def readTXT(filename, start_line=0, sep=None):
    with open(filename) as file:
        return [line.rstrip().split(sep) for line in file.readlines()[start_line:]]

## Song

In [6]:
songs = pd.read_csv('./data/bronze/song_hash.txt', sep = '\t', header = None,
                    names = ['song_id', 'title', 'artist'], index_col = 0)
songs['artist - title'] = songs['artist'] + " - " + songs['title']
songs.head()

title  ...                                     artist - title
song_id                                                     ...                                                   
0                             Gucci Time (w\/ Swizz Beatz)  ...          Gucci Mane - Gucci Time (w\/ Swizz Beatz)
1        Aston Martin Music (w\/ Drake & Chrisette Mich...  ...  Rick Ross - Aston Martin Music (w\/ Drake & Ch...
2                            Get Back Up (w\/ Chris Brown)  ...               T.I. - Get Back Up (w\/ Chris Brown)
3                       Hot Toddy (w\/ Jay-Z & Ester Dean)  ...         Usher - Hot Toddy (w\/ Jay-Z & Ester Dean)
4                                             Whip My Hair  ...                              Willow - Whip My Hair

[5 rows x 3 columns]

## Tag

In [14]:
tags = readTXT('./data/bronze/tags.txt')
tags[7:12]

[['49', '65', '72', '141', '197'], ['11', '35', '154'], ['#'], ['#'], ['#']]

> Note: # means the song doesn't have any tag. we can replace it with unknown

In [16]:
mapping_tags = dict(readTXT('./data/bronze/tag_hash.txt', sep = ', '))
mapping_tags['#'] = "unknown"
song_tags = pd.DataFrame({'tag_names': [list(map(lambda x: mapping_tags.get(x), t)) for t in tags]})
song_tags.index.name = 'song_id'
song_tags.head()

tag_names
song_id                                                   
0                                                [wjlb-fm]
1        [chill, rnb, loved, hip hop, rap, soft, wjlb-f...
2                                                [wjlb-fm]
3                                           [pop, hip-hop]
4        [pop, american, dance, rnb, hip-hop, hip hop, ...

We will consider song tags as a feature of song, so will merge it in songs dataset

In [17]:
songs = pd.merge(left = songs, right = song_tags, how = 'left',
                 left_index = True, right_index = True)
songs.index = songs.index.astype('str')
songs.head()

title  ...                                          tag_names
song_id                                                     ...                                                   
0                             Gucci Time (w\/ Swizz Beatz)  ...                                          [wjlb-fm]
1        Aston Martin Music (w\/ Drake & Chrisette Mich...  ...  [chill, rnb, loved, hip hop, rap, soft, wjlb-f...
2                            Get Back Up (w\/ Chris Brown)  ...                                          [wjlb-fm]
3                       Hot Toddy (w\/ Jay-Z & Ester Dean)  ...                                     [pop, hip-hop]
4                                             Whip My Hair  ...  [pop, american, dance, rnb, hip-hop, hip hop, ...

[5 rows x 4 columns]

We will remove the unknown songs, which doesn't have title and artist. 

In [18]:
unknown_songs = songs[(songs['artist'] == '-') | (songs['title'] == '-')]
songs.drop(unknown_songs.index, inplace = True)

## Playlist

In [19]:
playlist = readTXT('./data/bronze/train.txt', start_line = 2) + readTXT('./data/bronze/test.txt', start_line = 2)
print(f'Playlist Count: {len(playlist)}')

Playlist Count: 15910


In [20]:
for i in range(0, 3):
    print("-------------------------")
    print(f"Playlist Idx. {i}: {len(playlist[i])} Songs")
    print("-------------------------")
    print(playlist[i])

-------------------------
Playlist Idx. 0: 97 Songs
-------------------------
['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '2', '42', '43', '44', '45', '46', '47', '48', '20', '49', '8', '50', '51', '52', '53', '54', '55', '56', '57', '25', '58', '59', '60', '61', '62', '3', '63', '64', '65', '66', '46', '47', '67', '2', '48', '68', '69', '70', '57', '50', '71', '72', '53', '73', '25', '74', '59', '20', '46', '75', '76', '77', '59', '20', '43']
-------------------------
Playlist Idx. 1: 205 Songs
-------------------------
['78', '79', '80', '3', '62', '81', '14', '82', '48', '83', '84', '17', '85', '86', '87', '88', '74', '89', '90', '91', '4', '73', '62', '92', '17', '53', '59', '93', '94', '51', '50', '27', '95', '48', '96', '97', '98', '99', '100', '57', '101', '102', '25', '103', '3', '1

In [22]:
# Remove unknown songs from the playlist.
playlist_wo_unknown = [[song_id for song_id in p if song_id not in unknown_songs.index]
                       for p in playlist]

# Remove playlist with zero or one song, since the model wouldn't capture any sequence in that list.
clean_playlist = [p for p in playlist_wo_unknown if len(p) > 1]
print(f"Playlist Count After Cleansing: {len(clean_playlist)}")

# Remove song that doesn't exist in any playlist.
unique_songs = set(itertools.chain.from_iterable(clean_playlist))
song_id_not_exist = set(songs.index) - unique_songs
songs.drop(song_id_not_exist, inplace = True)
print(f"Unique Songs After Cleansing: {songs.shape[0]}")

Playlist Count After Cleansing: 15842
Unique Songs After Cleansing: 73448


Before there were 75262 unique songs and 15910 playlists. Now we are ready with 73448 unique songs and 15842 playlists.


## Save the artifacts

In [23]:
!mkdir ./data/silver

with open('./data/silver/songs.pickle', 'wb') as handle:
    pickle.dump(songs, handle, protocol=pickle.HIGHEST_PROTOCOL)

with open('./data/silver/clean_playlist.pickle', 'wb') as handle:
    pickle.dump(clean_playlist, handle, protocol=pickle.HIGHEST_PROTOCOL)